[RF-Docs] Accessibility guide (#57293)
rails/rails: guides/source/accessibility.md

* How to use native HTML for common patterns like disclosures, dialogs,
  popovers, and tooltips.
* How to write CSS that keeps focus, target size, contrast, zoom, and user
  preferences in check.
* How to manage focus, announce updates, signal in-flight submissions, and
  confirm destructive actions in apps driven by Turbo.
* How to test for accessibility with automated tools and with the same input
  methods real users rely on.

--------------------------------------------------------------------------------

Why Accessibility Matters
-------------------------

Accessibility means building applications that work for people regardless of how
they use a computer: keyboard only, screen reader, magnifier, voice control,
switch device, or any combination.

Most of the web does not. The [WebAIM Million][webaim-million] report audits the
top one million home pages every year and consistently finds detectable WCAG
failures on around 95% of them. The [World Health Organization
estimates][who-disability] that roughly one in six people lives with a
significant disability, so the gap between that audience and the inaccessible

> state of the web is large.

Reviewer: This section does a great job of describing the data and statistics related to disability.

> significant disability

In addition to what's described as "significant" disability (which I interpret to also mean "permanent"), could it be worth mentioning more temporary disabilities (having a broken arm or broken fingers, or having dilated eyes and impaired vision after an optometrist appointment, for example) or situational/environmental disabilities (being a parent holding a child and only being limited to browsing with one hand, browser in a dark environment, for example)?

Author: Yes, makes sense! Added a paragraph on temporary and situational disability.

### Assistive Technologies

An **assistive technology** is any tool that helps a person interact with a
computer. The categories below are the ones that matter most when building for
the web:

* **Screen readers:** Programs that convert screen content into speech or
  braille output. A blind or visually impaired user listens to a synthetic voice
  that announces text, headings, links, buttons, and form controls as they
  navigate. Screen readers do not "see" the page; they read the underlying HTML
  structure. Popular screen readers include [NVDA](https://www.nvaccess.org/)
  (Windows), [VoiceOver][vo-mac-guide] (built into macOS),
  [Orca](https://orca.gnome.org/) (Linux), [VoiceOver][vo-ios-guide] (built into
  iOS and iPadOS), and [TalkBack][talkback-guide] (built into Android).

* **Screen magnifiers:** Programs that enlarge portions of the screen for people
  with low vision. Users see only a small area of the page at a time and
  navigate by panning around. Good structure, consistent layouts, and clear
  focus indicators help magnifier users orient themselves. Windows includes
  Magnifier, macOS and iOS/iPadOS include Zoom, and Android includes

>   Magnification.

Reviewer 2: Maybe start with MacOS first, to clarify that we are talking about Windows the OS, not a browser window?
```suggestion
  focus indicators help magnifier users orient themselves. MacOS and iOS/iPadOS include
  Zoom, Windows includes Magnifier,  and Android includes
  Magnification.
```

Author: I had phrased it that way to maintain consistency with the guide, which always mentions Windows first, but if it makes things clearer, I guess that’s better. Reordered to lead with macOS.

* **Voice control:** Software that allows users to speak commands instead of
  using a keyboard or mouse. A user might say "click Submit" or "click Email" to
  interact with buttons and form fields. For this to work, every interactive

>   element needs a visible text label: if a button has no text, voice control

Reviewer 3: This feels a bit too absolute. Later the guide says icon-only buttons with an `aria-label` can be reached by voice control, so "every interactive element needs a visible text label" contradicts that.

Author: You're right, reworded so the icon-only button with an `aria-label` is still reachable, the real failure is a control with no accessible name at all.

  users cannot refer to it by name. Voice control is built into all major
  platforms: Voice Control on macOS/iOS, Voice Access on Windows and Android.

* **Switch devices:** Physical devices used by people with significant motor
  limitations. A switch device might be a single button or a sip-and-puff
  device. The OS scans the interface and highlights items in turn, and the user
  presses the switch to activate the current target. Keyboard accessibility and
  a clean focus order are essential for switch users. Switch Control is built
  into macOS, iOS, and iPadOS, and Switch Access is built into Android.

* **Hands-free pointing:** Head and eye trackers let people with significant
  motor limitations control a cursor without a mouse. Activation usually relies

>   on dwell-to-click, so adequate target size and pointer-friendly interactions

Reviewer 2: I'm not familiar with dwell-to-click. Maybe that should be explained here?

Author: Right, defined it inline.

them automatically depending on whether the user is reading or interacting with
a control.

##### Browse Mode

Browse mode is the default and is what the user stays in while reading. The
screen reader intercepts every keystroke and uses it as a navigation command, so
the keyboard effectively belongs to the tool rather than the browser.

This setup exists to enable **quick navigation** through the page. A sighted
user scans a page visually to spot the main article, a sidebar, navigation
links, or a search form, usually in a second or two. Browse mode gives a screen
reader user the same ability in audio: a single keystroke jumps to the next
heading on the page and announces it; another keystroke jumps to the next one;
another to the next landmark region; another to the next link. The user can
survey the page structure this way without listening to the paragraphs in
between, then stop at whatever they want to read in full.

The categories readers jump through are consistent everywhere: headings (and
specific heading levels), landmark regions, links, buttons, form fields, tables,
graphics, and lists. How each category is triggered differs per reader, though.

> NVDA and Orca assign one letter of the alphabet to each category: pressing that

Reviewer 2: Maybe remind readers here that NVDA and Orca are screen readers?
```suggestion
Screen readers NVDA and Orca assign one letter of the alphabet to each category: pressing that
```

Author: Done, reworded to lead with "Screen readers".

letter jumps to the next element of that type. VoiceOver on macOS calls the
feature the **Rotor**, a category picker the user opens to choose what to step
through, and also offers a single-letter mode called **Quick Nav** that behaves
like NVDA and Orca.

This is why HTML structure matters so much. Quick navigation only works when the
page has the structure it jumps through. If a page has no headings, the heading
shortcut does nothing. If a page has no landmark regions, the landmark shortcut
does nothing. The user is left to arrow through every single line of content to
find what they need, like reading an entire book that has no table of contents.

##### Focus Mode

Focus mode is the opposite. Every keystroke is passed through to the browser as

> if the screen reader were not running, so pressing `H` types the letter "h",

Reviewer 2: When I read "pressing `H`" I interpret that as pressing the Shift and h key.
Should this be "pressing `h`" (with a lower case) instead?

Reviewer 4: Maybe a good fit for the `<kbd>` tag

https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/kbd

Author: On the casing, I went back and forth. [NVDA's docs](https://download.nvaccess.org/documentation/userGuide.html) do write these lowercase, but they're lowercase across the board (`control`, `escape` too), which would clash with the capitalized modifiers used elsewhere in the guide. [Deque](https://dequeuniversity.com/screenreaders/nvda-keyboard-shortcuts), [Apple's VoiceOver docs](https://support.apple.com/guide/voiceover/general-commands-cpvokys01/mac), and the [Google](https://developers.google.com/style/ui-elements) and [Microsoft](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/term-collections/keys-keyboard-shortcuts) style guides all go uppercase letter with capitalized modifier. The `H`/`h` split is intentional too, the key is `H`, the letter it types is `h`. That said, happy to lowercase the single nav keys, I don't have a strong opinion here.

On `<kbd>`, agreed it's the proper element for keys. Doing it would mean wrapping every key across the command tables, and since the guides theme doesn't style `<kbd>` today (just a normalize reset), it would also want a little CSS added to the theme so the keys don't render unstyled. I'm open to doing it here. Do you think it belongs in this PR alongside the guide content, or is touching the theme styling cleaner as its own change?

of **semantic HTML** becomes clear. Before reaching for a `<div>`, check whether
HTML already has a dedicated element for the task. Semantic HTML means using
elements for their intended purpose: `<button>` for actions, `<a>` for links,
`<h1>`-`<h6>` for headings, `<nav>` for navigation, and so on.

Each semantic element carries built-in meaning, keyboard behavior, and screen
reader support. Consider the difference:

```html
<!-- Non-semantic: a <div> styled to look like a button -->
<div class="btn btn-primary">Save</div>

<!-- Semantic: an actual <button> element -->
<button type="submit" class="btn btn-primary">Save</button>
```

These may look identical on screen, but the `<div>` has none of the built-in
behavior that the `<button>` provides. A native `<button>`:

* Is focusable with the keyboard (Tab key).
* Can be activated with Enter or Space.
* Is announced as "button" by screen readers.
* Can be activated by voice control users saying "click Save."
* Submits its parent form when pressed.
* Shows a native focus indicator.

Some of this can be added manually: `tabindex="0"` for focus, `role="button"`

> for screen reader announcement, JavaScript handlers for Enter, Space, and click.

Reviewer 2: Nitpick. This sentence was a bit difficult to read for me.

Maybe make it a list as well? This could also make it easier to compare the native behavior with the manual behavior:
```suggestion
Some of this can be added manually:

* `tabindex="0"` for focus
* JavaScript handlers for Enter, Space, and click.
* `role="button"` for screen reader announcement

```

Author: Made it a list, parallel to the native-behavior list above.

But the browser provides many subtle behaviors for free, and it is easy to miss
one: form submission, `disabled` state management, focus handling, interaction

> with screen reader modes, and more. The native element is what guarantees the

Reviewer 2: Nitpick. I first read this as "form submission" being the easy one to miss: "But the browser provides many subtle behaviors for free, and it is easy to miss one: form submission."

Maybe first list the behaviors and move the "easy to miss" part to a new sentence?
```suggestion
But the browser provides many subtle behaviors for free: form submission, `disabled` state management, focus handling, interaction
with screen reader modes, and more. It is easy to miss one. The native element is what guarantees the
```

Author: Added the colon that introduces the behaviors, and "It is easy to miss one" is its own sentence.

referencing the `id` of another element whose text should serve as the name:

[aria-labelledby]: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-labelledby

```html+erb
<nav aria-labelledby="nav-heading">
  <h2 id="nav-heading">Product Categories</h2>
  <ul>
    <li><%= link_to "Electronics", electronics_path %></li>
    <li><%= link_to "Clothing", clothing_path %></li>
  </ul>
</nav>
```

The screen reader announces "Product Categories, navigation" when the user
reaches this landmark. Prefer `aria-labelledby` over `aria-label` when the
naming text is already visible on the page. It reuses existing content and keeps
a single source of truth, so if someone updates the heading text, the accessible
name updates automatically. With `aria-label`, it is common for someone to
change the visible text and forget to update the attribute, leaving the
accessible name out of sync and the visible label no longer included in the name
voice control users would say ([WCAG 2.5.3 Label in Name][wcag-label-in-name]).

[wcag-label-in-name]: https://www.w3.org/WAI/WCAG22/Understanding/label-in-name.html

> The `aria-labelledby` attribute accepts more than one ID separated by spaces.

Reviewer 2: Nitpick. Adding a comma would make it clearer "separated by spaces" applies to "more than one ID" instead of "one ID seperated by spaces":
```suggestion
The `aria-labelledby` attribute accepts more than one ID, separated by spaces.
```

Author: Added it, and the same in the parallel `aria-describedby` sentence.

This lets the accessible name compose from several existing pieces of page
content without duplicating strings.

> 

Reviewer 2: There are several paragraphs here, discussing different topics, without any headers.

Maybe a header could be added to the paragraphs describing some naming best practices?
For example:
```suggestion

#### Naming Best Practices

```

Author: Added the "Naming Best Practices" heading.

only works if the word "Save" appears in the accessible name. An
`aria-label="Submit form"` on a button whose visible text reads "Save" is a
silent bug for those users, even though a screen reader announces it cleanly.
When in doubt, use `aria-labelledby` pointing at the visible text and let the
browser keep the two in sync.

The accessible name should name the control, not instruct the user how to
operate it. Screen readers already announce the role and how each control is
activated, so an `aria-label="Click button to mark"` or a `<legend
aria-label="Use arrow keys to choose one">` adds noise that conflicts with what
the screen reader says natively. The instruction often does not even apply,
since the user may be on touch or voice control rather than the input method the
hint assumes. The same principle covers pronunciation. Phonetic spellings or
splitting a word across `<span>` letters to make the reader say a name
differently break translation, find-in-page, and braille output, and they
conflict with how the user has configured their reader. Name the control, trust
the native announcement, and rewrite the name itself if it causes friction in
pronunciation.

> NOTE: The HTML [`title`][title-attr] attribute is sometimes used for the same

Reviewer 2: I think this note about the title attribute could be mentioned a bit earlier.
Maybe right after the first example of adding the label to the icon-only button?

Author: Moved it up next to that example.

only reveal it as a tooltip on mouse hover, so touch and keyboard users cannot
see it, and screen reader support varies. It can be useful as an additional hint
on an element that already has a visible label or an `aria-label`, but do not
rely on it to name a control.

[title-attr]: https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Global_attributes/title

#### Naming Limitations and Visually Hidden Text

`aria-label` and `aria-labelledby` only work on elements whose **role** supports
naming: interactive elements, landmarks, and other identifiable roles like
`<img>`, `<table>`, `<dialog>`, `group`, or `list`. They are **prohibited** on
elements with a role that does not support naming. A plain `<div>`, `<span>`, or
`<p>` has a role (`generic` or `paragraph`) that prohibits naming, so
`aria-label` on these elements is invalid and screen readers will ignore it.
However, the restriction is on the *role*, not the *element*: `<div role="group"
aria-label="...">` is valid because the `group` role supports naming, even
though the underlying element is a `<div>`. The WAI-ARIA specification lists all
[roles supporting name from

> author](https://www.w3.org/TR/wai-aria-1.2/#namefromauthor).

Reviewer 2: The link "roles supporting name from author" is the first time "author" is mentioned.

In context of the paragraph, would it be clearer if we replace "roles supporting name from
author" with "roles supporting naming" for the link?

```suggestion
though the underlying element is a `<div>`. The WAI-ARIA specification lists all
[roles supporting naming](https://www.w3.org/TR/wai-aria-1.2/#namefromauthor).
```

Author: Changed the link text to "roles supporting naming", same target.

This means that on a generic element like a paragraph or a `<div>`, `aria-label`
is ignored entirely. And on elements whose role does accept it, the attribute
*replaces* the accessible name rather than adding to it.

> The solution is **visually hidden text**, a `<span>` that is visible to screen

Reviewer 2: It's a bit unclear from reading the text which problem the following solution solves.
What would be a reason to still keep the notifications in spans?
Also, would it help to show an example with problematic code?

For example (not sure if this is correct):
````suggestion
```html+erb
<%# the aria-label is ignored here %>
<span>🔔</span>
<span aria-label="unread notifications">3</span>
```

The solution is **visually hidden text**, a `<span>` that is visible to screen
````

Author: True, added a "before" example and made the failure explicit: `aria-label` on a generic element is invalid and announced inconsistently.

contact", giving the user context about what the button does without changing
its name.

[`aria-describedby`][aria-describedby] references the `id` of another element
that contains the descriptive text. Use it when the description is already
visible on the page, such as a hint below a form field:

[aria-describedby]: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-describedby

```html+erb
<%= form.label :password %>
<%= form.password_field :password, aria: { describedby: "password-hint" } %>
<p id="password-hint">Must be at least 12 characters.</p>
```

The screen reader announces the label, the field type, and the description, for
example: "Password, secure text field, Must be at least 12 characters."

Like `aria-labelledby`, `aria-describedby` accepts more than one ID separated by
spaces, so a description can compose from several existing pieces of page
content without duplicating strings.

NOTE: Like `aria-label`, descriptions only work reliably on elements whose role
supports them: interactive elements, landmarks, and other identifiable roles.
They will not work on a plain `<div>`, `<span>`, or `<p>`, unless the element

> has an explicit role that supports naming.

Reviewer 3: Small wording issue: this says descriptions work on roles that support "naming", but this section is about descriptions. Should this be "roles that support descriptions" instead?

Author: Good catch! Since it works similarly to a label in terms of consistency in screen readers, I assumed the same rule applied, but neither was right. `aria-describedby` and `aria-description` are global, not role-restricted like `aria-label`. Reworded so descriptions are valid on any element but announced reliably only on an interactive control.

    <h3><%= slide.title %></h3>
    <%= link_to "Read more", slide.url, tabindex: (index == @active_index ? nil : -1) %>
  <% end %>
<% end %>
```

With one link per slide this is tolerable; with two or three focusable elements
per slide, and a template that evolves over time, it becomes easy to miss one.

The [`inert`][inert-attr] attribute handles the whole subtree in one
declaration. It removes the content from the accessibility tree, prevents focus
on any descendant, **and** blocks click events inside. Moving `inert` between
the active and inactive slides is enough to keep the carousel accessible as it
rotates, with no `tabindex` bookkeeping:

[inert-attr]: https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Global_attributes/inert

```html+erb
<% @slides.each_with_index do |slide, index| %>
  <%= tag.div class: "slide", inert: index != @active_index do %>
    <%= image_tag slide.image, alt: slide.alt %>
    <h3><%= slide.title %></h3>
    <%= link_to "Read more", slide.url %>
  <% end %>
<% end %>
```

Rails omits the attribute when the value is `false`, so the active slide renders
without `inert` and every other slide carries it. The single attribute covers

> every descendant, including ones added to the template later.

Reviewer 2: Great example of how this applies to Rails!

see which page they are on. But a screen reader user hears a list of links that
all sound the same. Nothing tells them which one represents the current page.

[`aria-current="page"`][aria-current] solves this. The screen reader announces
"current page" alongside the link text:

[aria-current]: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-current

```html+erb
<nav>
  <ul>
    <li>
      <%= link_to "Home", root_path, aria: { current: ("page" if current_page?(root_path)) } %>
    </li>
    <li>
      <%= link_to "Articles", articles_path,
            aria: { current: ("page" if current_page?(articles_path)) } %>
    </li>
  </ul>
</nav>
```

The same attribute can also be used as a CSS selector, making it the single
source of truth for both the visual style and the screen reader announcement:

```css
[aria-current="page"] {
  font-weight: bold;
  border-bottom: 2px solid currentColor;
}
```

`aria-current` accepts other values for contexts where "page" does not fit:
`step` for a multi-step flow, `location` for the active item in a map or tree,
`date` for a calendar's selected day, `time` for a timeline, or `true` as a

> generic fallback.

Reviewer: There is currently an open PR to change `link_to` helper calls to behave like the example in the code snippet:

https://github.com/rails/rails/pull/57350

Insights like this paragraph may be valuable in reviewing that proposed code change.

If that PR is merged (either before or after this change to the Guides), would it make sense to mention the new behavior in the code sample?

Author: It was interesting, and it was great to have inspired that PR with the guide, but #57350 was closed.

[`aria-pressed`][aria-pressed] tells the screen reader the button's current
state. It announces "pressed" or "not pressed" along with the button's name,
which is how the toggle state becomes programmatically available ([WCAG 4.1.2
Name, Role, Value][wcag-name-role-value]). A [Stimulus
controller][stimulus-controllers] can manage this, using a value so the initial
state can be set from the server:

[aria-pressed]: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-pressed
[stimulus-controllers]: https://stimulus.hotwired.dev/reference/controllers

```js
// app/javascript/controllers/toggle_button_controller.js
import { Controller } from "@hotwired/stimulus"

export default class extends Controller {
  static values = { pressed: Boolean }

  pressedValueChanged() {
    this.element.ariaPressed = this.pressedValue
  }

  press() {
    this.pressedValue = !this.pressedValue
  }
}
```

```html+erb
<%= tag.button "Like", type: "button",
      data: {
        controller: "toggle-button",
        toggle_button_pressed_value: @article.liked_by?(current_user),
        action: "toggle-button#press" } %>

> ```

Reviewer: Maybe I'm misunderstanding the guidance being provided here, but could the sample be simplified by replacing the [Stimulus Value](https://stimulus.hotwired.dev/reference/values#change-callbacks) usage with directly declaring `aria: { pressed: … }` and reading from and writing to [ariaPressed](https://developer.mozilla.org/en-US/docs/Web/API/Element/ariaPressed)?

```js
// app/javascript/controllers/toggle_button_controller.js
import { Controller } from "@hotwired/stimulus"

export default class extends Controller {
  press() {
    this.element.ariaPressed = !this.element.ariaPressed
  }
}
```

```html+erb
<%= tag.button "Like", type: "button",
      aria: { pressed: @article.liked_by?(current_user) },
      data: { controller: "toggle-button", action: "toggle-button#press" } %>
```

Author: I did it this way because of Stimulus's automatic type casting of values, ARIA properties return a string even though they are actually booleans, but it's true that it's not really necessary, and it's simpler this other way. As `element.ariaPressed` is a string, so `!this.element.ariaPressed` is always false (a non-empty string is truthy) and the button never toggles back, I went with `this.element.ariaPressed = this.element.ariaPressed !== "true"`.

When using `aria-pressed`, do not change the button's text to reflect the state.
If the button says "Like" when unpressed and "Unlike" when pressed, the screen
reader announces "Unlike, pressed", which contradicts itself. Keep the text

> stable and let `aria-pressed` carry the state.

Reviewer 2: After reading this paragraph I wondered what would be the solution if I do want to change the button's text. It's only 2 paragraphs later that it is explained.
I wonder if that paragraph should follow this one, or maybe the paragraphs can be combined?

Author: Added a pointer to the `aria-live` version right where the question comes up.

Also avoid toggle verbs like "Toggle sidebar" in the label: saying "toggle"
duplicates what `aria-pressed` already communicates. Prefer action verbs like
"Show sidebar": the screen reader announces "Show sidebar, button, pressed" or

> "Show sidebar, button, not pressed", which is clear and not redundant.

Reviewer: This is valuable guidance!

When the design changes the button text on toggle (for example, "Show completed"
and "Hide completed") instead of carrying `aria-pressed`, the label itself is
what indicates the state. [Announcing Dynamic Changes with
`aria-live`](#announcing-dynamic-changes-with-aria-live) covers how to make that
change reach the screen reader.

TIP: For settings or preferences that toggle between on and off, consider using
a native `<input type="checkbox" switch>` instead. It renders as a toggle switch
and the screen reader announces its on/off state automatically. In browsers that

> do not yet support the `switch` attribute, it falls back to a regular checkbox.

Author: See https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/input#switch

[role-status]: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Roles/status_role
[role-alert]: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Roles/alert_role
[aria-atomic]: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-atomic
[aria-relevant]: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-relevant
[wcag-status-messages]: https://www.w3.org/WAI/WCAG22/Understanding/status-messages.html

There are two levels of urgency:

* **`aria-live="polite"`** (or [`role="status"`][role-status]): The screen
  reader waits until the user is idle before announcing. Use for non-urgent
  updates.
* **`aria-live="assertive"`** (or [`role="alert"`][role-alert]): The screen
  reader interrupts immediately. Use sparingly, for errors and critical warnings
  only.

Two related attributes control what gets announced:

* [`aria-atomic="true"`][aria-atomic]: Announces the **entire** live region
  content when any part changes, not just the changed part. Useful when the
  change only makes sense in context, for example a clock showing "12:35 PM"

>   where announcing just "35" would be meaningless.

Reviewer 2: Maybe split this sentence and make it explicit what will be announced?
```suggestion
  change only makes sense in context. For example, a clock changing to "12:35 PM"
  should announce "12:35 PM" not just "35".
```

Author: Good point, changed to: should announce "12:35 PM", not just "35".

  The default is `additions text`, which is almost always what is needed.
  Override it only when a removal is genuinely meaningful on its own.

For example, a button that changes its visible text to reflect the current state
needs a live region so screen readers announce the new label. A Stimulus
controller manages the state with a value, and CSS shows the appropriate text:

```js
// app/javascript/controllers/swap_controller.js
import { Controller } from "@hotwired/stimulus"

export default class extends Controller {
  static values = { pressed: Boolean }

  toggle() {
    this.pressedValue = !this.pressedValue
  }
}
```

```html+erb
<%= tag.button type: "button",
      data: {
        controller: "swap",
        swap_pressed_value: @showing_completed,
        action: "swap#toggle" } do %>
  <span aria-live="polite" aria-atomic="true" aria-relevant="all">
    <span data-swap-target="primary">Show completed</span>
    <span data-swap-target="secondary">Hide completed</span>
  </span>
<% end %>
```

```css
[data-swap-pressed-value="false"] [data-swap-target="secondary"],
[data-swap-pressed-value="true"] [data-swap-target="primary"] {
  display: none;
}

> ```

Reviewer 2: While looking at the example, I couldn't understand how this example worked because the css was outside my browser window. Maybe move the css code example between the js and html code examples?

Reviewer: Would it make sense to declare *both* in the `html+erb` block by nesting the CSS in a `<style>` element?

```html+erb
<style>
  [data-swap-pressed-value="false"] [data-swap-target="secondary"],
  [data-swap-pressed-value="true"] [data-swap-target="primary"] {
    display: none;
  }
</style>

<%= tag.button type: "button",
      data: {
        controller: "swap",
        swap_pressed_value: @showing_completed,
        action: "swap#toggle" } do %>
  <span aria-live="polite" aria-atomic="true" aria-relevant="all">
    <span data-swap-target="primary">Show completed</span>
    <span data-swap-target="secondary">Hide completed</span>
  </span>
<% end %>
```

Author: Reordered to HTML, CSS, JS so the CSS sits next to its markup. I kept them as separate blocks rather than an inline `<style>` to match how CSS is shown elsewhere in the guide. I'm not sure if we should display it as a style element, given CSP and what's generally recommended in Rails, but I'm open to it if you prefer that.

works in `link_to`, `button_to`, `form_with`, `image_tag`, and all other view
helpers.

When generating HTML dynamically, the [tag helper][tag-helper] accepts the same
`aria:` option.

[tag-helper]: https://api.rubyonrails.org/classes/ActionView/Helpers/TagHelper.html#method-i-tag

A common pattern is iterating over a collection and setting ARIA attributes
conditionally per item. For example, a checkout progress indicator marks the
step the user is currently on:

```html+erb
<nav aria-label="Checkout progress">
  <ol>
    <% @checkout_steps.each do |step| %>
      <%= tag.li aria: { current: ("step" if step == @current_step) } do %>
        <%= link_to step.name, step.path %>
      <% end %>
    <% end %>
  </ol>
</nav>
```

Each `<li>` evaluates `step == @current_step`: the one that matches gets
`aria-current="step"`, the rest get nothing. Rails skips ARIA attributes whose
value is `nil`, but `false` renders as the literal string `"false"` instead, so
reach for `nil` when the attribute should not appear at all. The screen reader
announces "current step" on the active item, so users know where they are in the
flow even when the visual styling alone conveys it.

> Helpers also normalize array values. For ARIA attributes that take a
> space-separated list of tokens, like `aria-labelledby` and `aria-describedby`,
> passing an array joins the entries with a single space and drops `nil` or
> `false` entries. This keeps references to other elements declarative even when
> the list depends on the state of the record.

Reviewer 2: I think passing an array to any ARIA attribute joins the entry as an array:
https://github.com/rails/rails/blob/ad36942b04e9c31ccee5ea4b0071944d30fcf43e/actionview/lib/action_view/helpers/tag_helper.rb#L1532

Maybe we could also show examples here?
````suggestion
Helpers also normalize array values. For ARIA attributes that take a
space-separated list of tokens, like `aria-labelledby` and `aria-describedby`,
passing an array joins the entries with a single space and drops `nil` or
`false` entries. This keeps references to other elements declarative even when
the list depends on the state of the record.

```ruby
tag.span aria: { labelledby: ["foo", false, "bar"] }
=> "<span aria-labelledby=\"foo bar\"></span>"
```
````

Reviewer: The [token_list](https://edgeapi.rubyonrails.org/classes/ActionView/Helpers/TagHelper.html#method-i-token_list) view helper is also useful in situations that require a list of identifiers for `aria-` attributes:

```erb
<span aria-labelledby="<%= token_list("foo", false, "bar") %>"></span>
<%# => <span aria-labelledby="foo bar"></span> %>
```

Author: Added an example of the `aria:` option joining an array and dropping falsy entries, scoped to token-list attributes where an array is meaningful. I used the `aria:` form rather than `token_list`, which reads cleanest in the helper context, but `token_list` is a good tool for building the value directly. Want both shown? Since Rails already handles this automatically with an array, I thought it would be cleaner and more elegant to leave it that way.

> ### The First Rule of ARIA

Reviewer 2: If this is the first rule, shouldn't we mention it earlier? 😉 
Maybe move this section to the top?

Author: I thought about it, but it seemed odd to start talking about the first ARIA rule without first explaining what it does or why it’s there. I figured the rule would make more sense once you have a better understanding of what ARIA offers and how you might use it to bypass that rule. "The First Rule of ARIA" is the W3C's own term (the first of their [Rules of ARIA Use](https://www.w3.org/TR/using-aria/#firstrule)), so "first" is their numbering, not necessarily its position here. I added a link to that rule anyway, what do you think?

[link_to]: https://api.rubyonrails.org/classes/ActionView/Helpers/UrlHelper.html#method-i-link_to
[button_to]: https://api.rubyonrails.org/classes/ActionView/Helpers/UrlHelper.html#method-i-button_to

```html+erb
<%# Navigation: use a link %>
<%= link_to "View profile", profile_path(@user) %>

<%# Action: use a button %>
<%= button_to "Publish", article_publication_path(@article) %>
```

Two patterns commonly blur this line:

* **`<a href="#">` with a JavaScript handler attached.** The `#` is a
  placeholder, since the anchor points nowhere. Screen readers announce the
  element as a link to the same page, so the user expects to be scrolled to
  another section. Landing there and finding that the link performs an action,
  or does nothing at all, contradicts what they were told. Navigating to `#`
  also pushes an empty entry into browser history, so the back button stops
  working as expected. If the control performs an action, use a `<button>` (or
  `button_to`) from the start.

  ```html+erb
  <%# Avoid: an anchor with no destination, wired up with JavaScript %>
  <a href="#" data-action="notification#markAsRead:prevent">Mark as read</a>

  <%# Prefer: a button %>

>   <%= button_to "Mark as read", notification_readings_path(@notification) %>

Reviewer: Based on guidance from MDN, a `<button>` element with `[type="button"]`:

> has no default behavior, and does nothing when pressed by default. It can have client-side scripts listen to the element's events, which are triggered when the events occur.

Would it make sense for this guide to mention that behavior-only `<button>` elements should be declared with [type="button"](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/button#type) so that they do not submit any `<form>` element ancestors when clicked?

Author: Yes I think so, added a note on it.

  ```

* **A link for an action that changes server state**, for example `link_to` with
  `data-turbo-method`. Even when the `href` is valid, the semantics are wrong. A
  user hearing "Publish article, link" reasonably expects to navigate to a page
  where they can review the article before submitting. Triggering the action
  directly on activation removes the review they were counting on. Prefer

>   `button_to`:

Reviewer: While it might not apply directly to the scenario outlined below, is it also worth mentioning that there is a difference between `<button>` and `<a>` elements when driving with a keyboard? 

An `<a>` element with focus will not respond to a <kbd>space</kbd> in the same way that it responds to a <kbd>enter</kbd> (i.e. triggering a "click"), but a `<button>` element will respond to both in the same way (i.e. triggering a "click").

Author: That's right! It's an important distinction, I forgot to mention it. Wrote that in.

  <%= link_to "Publish", article_publication_path(@article),
        data: { turbo_method: :post } %>

  <%# Avoid: a link that immediately deletes the article %>
  <%= link_to "Delete", article_path(@article),
        data: { turbo_method: :delete } %>

  <%# Prefer: a button that communicates the correct intent %>
  <%= button_to "Publish", article_publication_path(@article) %>
  <%= button_to "Delete", article_path(@article), method: :delete %>
  ```

A useful thought experiment applies to both anti-patterns. Every browser on
every platform lets the user open a link in a new tab or window. Does it make
sense to open "Publish article" or "Delete" in a new tab? Clearly not. Only
`GET` requests are safe to repeat or open out of context, which is exactly why
actions that change server state should not be links in the first place.

### Buttons inside Other Forms

A common reason developers use `link_to` with `data-turbo-method` is to avoid
nesting forms, for example a delete button inside a form that edits a record.
HTML does not allow nested `<form>` elements, and `button_to` generates its own
`<form>`.

The solution is the HTML [`form`][form-attr] attribute, which associates a

> button with a form **anywhere** in the DOM by referencing its `id`:

Reviewer: Browsers also support `<button>` and `<input type="button">` elements with [formmethod](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/button#formmethod) and [formaction](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/button#formaction) attributes.

In this scenario, browsers would ignore a `[formmethod="delete"] attribute on a submit button, but `@hotwired/turbo-rails` [maps a `[formmethod]` attribute to Rails' `_method` attribute prior to submission](https://github.com/hotwired/turbo-rails/pull/239).

Similarly, Action View supports [encoding `:formmethod` options to the `_method` parameter for HTTP verbs other than `GET` and `POST`](https://github.com/rails/rails/pull/41045).

Author: Didn't notice turbo-rails mapped `formmethod` to `_method`, nice, I remember having some issues with it in the past, but I never properly debugged it and just switched to the form element. Do you think it's worth adding this here to expand on the concept? In that case, we would accept that certain parameters are shared, like parameters that might not necessarily need to be sent, but it’s true that this approach is more elegant in some scenarios.

```html+erb
<%# Avoid: the link text carries no information on its own %>
To read our terms and conditions, <%= link_to "click here", terms_path %>.

<%# Prefer: the link text describes where it leads %>
<%= link_to "Read our terms and conditions", terms_path %>
```

Links that open in a new browser tab or window deserve extra care. The change of
context happens without warning, and keyboard users may lose their place when
focus moves to a new tab. When the behavior is necessary, signal it in the link
text, and include `rel="noopener"` to prevent the new page from gaining access
to the opener window:

```html+erb
<%= link_to "Read the WCAG specification (opens in a new tab)",
      "https://www.w3.org/TR/WCAG22/", target: "_blank", rel: "noopener" %>
```

### Disabled Links

A link with no destination is not really a link. HTML does not have a `disabled`
attribute for `<a>`: applying it has no effect in browsers, and screen readers
ignore it.

When an action that is normally a link should be unavailable, the simplest
answer is to not render a link at all. The classic example is pagination, where
the next control disappears at the end of the list:

```html+erb

> <%= link_to "Next", url_for(page: @page.next_param) unless @page.last? %>

Reviewer: While this example fits neatly on a single line, I think that my Ruby-parsing brain misinterpreted to be equivalent to calling `link_to "Next", nil`:

```html+erb
<%= link_to "Next", @page.last? ? url_for(page: @page.next_param) : nil %>
```

What do you think about changing the structure of the example so that it is more self-evident that the conditional skips rendering the element entirely?

```suggestion
<% unless @page.last? %>
  <%= link_to "Next", url_for(page: @page.next_param) %>
<% end %>
```

Author: Wrapped it in an explicit block.

what it refers to, with no table navigation, no header announcements, and no way
to jump between rows or columns.

Use the `<table>` element for tabular data:

```html+erb
<table>
  <caption>Monthly revenue by region</caption>
  <thead>
    <tr>
      <th scope="col">Region</th>
      <th scope="col">January</th>
      <th scope="col">February</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th scope="row">North</th>
      <td>$10,000</td>
      <td>$12,000</td>
    </tr>
  </tbody>
</table>
```

[`<caption>`][caption-el] gives the table a descriptive title that screen
readers announce when the user navigates to the table. `<th>` with
[`scope="col"`][scope-attr] marks column headers and `scope="row"` marks row
headers. These are what the screen reader reads aloud as the user moves between
cells. The navigation method varies by platform: keyboard shortcuts on desktop
(for example, `Ctrl + Alt + Arrow` keys in NVDA), or gestures on mobile (for
example, in VoiceOver on iOS, swipe left/right moves between columns and the
Rotor set to rows lets users swipe up/down to move between rows).

[caption-el]: https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/caption

> [scope-attr]: https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/th#scope

Reviewer: From a testing perspective, there's benefit to using elements like `<table>` that have built in semantics. For example, [Capybara](https://github.com/teamcapybara/capybara) has built-in support for a [`:table` selector](https://rubydoc.info/github/teamcapybara/capybara/master/Capybara/Selector) (as well as a [within_table](https://rubydoc.info/github/teamcapybara/capybara/master/Capybara/Session#within_table-instance_method) helper method) that knows how to resolve a table based on its rows, columns, and caption:

```ruby
page.find :table, "Monthly revenue by region"

assert_table "Monthly revenue by region", with_rows: [
  { "Region" => "North", "January" => "$10,000", "February" => "$12,000" }
]

within_table "Monthly revenue by region" do
  # …
end
```

Author: Nice extra reason to reach for a real `<table>`. Added a short note in the Tables section making that point, linking down to the system-tests section with the `within_table` example.

If the design truly cannot use a `<table>` element (for example, when the layout
requires CSS Grid or Flexbox), ARIA roles can replicate table semantics. This
preserves screen reader table navigation on non-table elements:

```html
<div role="table" aria-label="Monthly revenue by region">
  <div role="rowgroup">
    <div role="row">
      <div role="columnheader">Region</div>
      <div role="columnheader">January</div>
      <div role="columnheader">February</div>
    </div>
  </div>
  <div role="rowgroup">
    <div role="row">
      <div role="rowheader">North</div>
      <div role="cell">$10,000</div>
      <div role="cell">$12,000</div>
    </div>
  </div>
</div>
```

This gives screen readers the same structure as a native `<table>`, but it
requires significantly more markup and is easier to get wrong. Prefer a native
`<table>` whenever possible.

### Sortable Tables

When a table can be sorted by its column headers, sighted users see a visual
indicator (an arrow, bold text) showing which column is sorted and in which
direction. A screen reader user has no way to perceive these visual cues.

The [`aria-sort`][aria-sort] attribute on the `<th>` communicates the current

> sort state programmatically. It accepts `ascending`, `descending`, or `none`.

Reviewer 5: ```md
It accepts `ascending`, `descending`, or `none`.
```

I think `other` could also be a valid value—is there a specific reason why it’s been excluded?

Author: Yes, you're right. I don't remember ever needing it in practice, but there's no reason to leave it out, I've added it.

controller receives an array of IDs on submission, and `nil` as the unchecked
value keeps unselected rows out of that array.

[form-checkbox]: https://api.rubyonrails.org/classes/ActionView/Helpers/FormBuilder.html#method-i-checkbox

Each checkbox borrows its accessible name from the title cell next to it through
`aria-labelledby`, which gives the cell an `id` and lets the checkbox reuse that
visible text without duplicating it. The screen reader announces "The Rails Way,
checkbox" instead of an unlabeled "checkbox", and voice control users can
activate any row by saying the title they see on screen.

A Stimulus controller wires the select-all checkbox to the row checkboxes and
keeps both sides in sync. Toggling the `selectAll` target checks every `item`,
and changing any `item` checks the `selectAll` target when every item is
checked, clears it when none are, and sets it to `indeterminate` (the "mixed"
state assistive technology announces) when only some are:

```js
// app/javascript/controllers/selection_controller.js
import { Controller } from "@hotwired/stimulus"

export default class extends Controller {

>   static targets = [ "selectAll", "item" ]
> 
>   connect() {
>     this.refresh()
>   }
> 
>   toggleAll() {
>     this.itemTargets.forEach(item => item.checked = this.selectAllTarget.checked)

Reviewer: In an effort to simplify this Stimulus controller example code, could the `selectAll` target be replaced with an event listener that accesses the `<input type="checkbox">` through the Event's [target](https://developer.mozilla.org/en-US/docs/Web/API/Event/target) property?

```suggestion
  static targets = [ "item" ]

  connect() {
    this.refresh()
  }

  toggleAll(event) {
    this.itemTargets.forEach(item => item.checked = event.target.checked)
```

Then, the `selectAll` checkbox can drop the target:

```diff
- <%= form.checkbox :select_all,
-       data: { selection_target: "selectAll", action: "selection#toggleAll" } %>
+ <%= form.checkbox :select_all, data: { action: "selection#toggleAll" } %>
```

Author: I think the target has to stay here. `refresh()` needs a handle on the select-all checkbox to set its `checked`/`indeterminate` when items change (and on `connect`), and there the event target is an item, not the select-all. So `event.target` in `toggleAll` alone wouldn't let me drop it. Am I missing a way around that?

### Labels

Every form control must have a **label**, a visible piece of text that
identifies what it is for ([WCAG 3.3.2 Labels or
Instructions][wcag-labels-instructions]). Screen readers announce it on focus.
Without one, a screen reader user hears only the field type (for example, "edit
text, blank") with no indication of what to type. Voice control users cannot
refer to the field by name.

[wcag-labels-instructions]: https://www.w3.org/WAI/WCAG22/Understanding/labels-or-instructions.html

Rails form helpers automatically generate the `<label>` and `<input>` pairing:

```html+erb
<%= form_with model: @user do |form| %>
  <%= form.label :name %>
  <%= form.text_field :name %>

  <%= form.label :email_address %>
  <%= form.email_field :email_address %>

  <%= form.submit %>
<% end %>
```

This generates HTML where each `<label>` has a `for` attribute matching the
`<input>`'s `id`:

```html
<label for="user_name">Name</label>
<input type="text" name="user[name]" id="user_name">
```

This explicit association has multiple benefits: screen readers announce "Name"
when the input is focused, voice control users can say "click Name" to focus the

> field, and clicking the label focuses the input (enlarging the clickable area).

Reviewer: Again, System tests (and other HTML-specific tests) that utilize [Capybara](https://github.com/teamcapybara/capybara) can access a variety of [built-in selectors](https://rubydoc.info/github/teamcapybara/capybara/master/Capybara/Selector) like `:field` (or `:fillable_field` used with calls to [fill_in](https://rubydoc.info/github/teamcapybara/capybara/master/Capybara/Node/Actions#fill_in-instance_method)), `:checkbox` and `:radio_button` (used with calls to [check](https://rubydoc.info/github/teamcapybara/capybara/master/Capybara/Node/Actions#check-instance_method) / [uncheck](https://rubydoc.info/github/teamcapybara/capybara/master/Capybara/Node/Actions#uncheck-instance_method) and [choose](https://rubydoc.info/github/teamcapybara/capybara/master/Capybara/Node/Actions#choose-instance_method), respectively), and `:fieldset` (used with calls to [within_fieldset](https://rubydoc.info/github/teamcapybara/capybara/master/Capybara/Session#within_fieldset-instance_method)).

All of these selectors (and the assertions and actions that rely on them) are capable of resolving elements by their accessible names (`<label>` for form controls) and descriptions (`<fieldset>`-`<legend>` pairings), rather than CSS selectors. 

Strong Capybara conventions can drive more accessible user interfaces. Is that worth mentioning briefly here, with a link to the `#writing-accessible-system-tests` heading down below?

Author: Added a brief note in the Labels section that Capybara finds controls by their label (and fieldsets by their legend), so good labels make for simpler system tests, with a link down to Writing Accessible System Tests.

[axe-api]: https://www.deque.com/axe/core-documentation/api-documentation/

Use these options to push the conformance level higher than the default. For
example, enabling AAA contrast on top of the default rules:

```ruby
test "home page enforces AAA color contrast" do
  visit root_path
  assert_accessible checking: "color-contrast-enhanced"
end
```

Or running an audit against a specific set of standards. Note that
`according_to:` replaces the default rule set, so every tag that should apply
must be listed explicitly. The example below covers WCAG 2.2 AA plus
best-practice, Section 508 (United States), and EN 301 549 (European Union):

```ruby
test "home page conforms to WCAG 2.2 AA plus Section 508 and EN 301 549" do
  visit root_path
  assert_accessible according_to: [
    "wcag2a", "wcag2aa", "wcag21a", "wcag21aa", "wcag22aa",
    "best-practice", "section508", "EN-301-549"
  ]
end
```

#### Writing Accessible System Tests

Write system tests that interact with the application the way assistive
technology users do. [Rails system tests](testing.html#system-testing) use
[Capybara][capybara] under the hood, and its finders already work in terms of

> the accessible name. Prefer finding elements by their **accessible names**

Reviewer: What do you think about replacing "finders" with "selectors", since a [selector](https://rubydoc.info/github/teamcapybara/capybara/master/Capybara/Selector) is a first-party term used by Capybara?

Would it be worth linking to the list of [first-party Selectors](https://rubydoc.info/github/teamcapybara/capybara/master/Capybara/Selector) that Capybara provides?

Author: Yes, switched to "selectors" and linked the list.

fill_in "Email address", with: "user@example.com"
```

Use semantic finders for actions, and their `assert_*` counterparts for
verification. Both locate elements by accessible name:

```ruby
click_button "Save"            # finds by accessible name
click_link "View profile"      # finds by link text
check "Remember me"            # finds by label text
choose "Monthly billing"       # finds by label text
select "Canada", from: "Country"

assert_button "Save"           # asserts a button with that name exists
assert_link "View profile"     # asserts a link with that text exists
assert_field "Email address", with: "user@example.com"
```

Scope assertions within landmarks so the test verifies page structure and not
just the presence of text anywhere on the page. When a page has multiple
landmarks of the same type, resolve them by role and accessible name:

```ruby
within "main" do
  assert_text "Welcome"
end

# When several <nav> elements exist, disambiguate by their accessible name:
within :element, "nav", "aria-label" => "Primary" do
  assert_link "Home", href: root_path
end

within :element, "nav", "aria-label" => "Footer" do
  assert_link "Privacy", href: privacy_path
end

> ```

Reviewer: The [capybara_accessible_selectors](https://github.com/citizensadvice/capybara_accessible_selectors) gem is a third-party extension to Capybara's built-in selectors.

It provides selectors for a variety of accessibility concepts, including landmarks:

```ruby
# https://github.com/citizensadvice/capybara_accessible_selectors#main
within :main, "Welcome" do # "Welcome" here is the locator argument, which resolves to an <h1>
 # …
end

https://github.com/citizensadvice/capybara_accessible_selectors#navigation
# When several <nav> elements exist, disambiguate by their accessible name:
within :navigation, "Primary" do # "Primary" is the locator argument, which resolves [aria-label] or [aria-labelledby]
  assert_link "Home", href: root_path
end

within :navigation, "Footer" do
  assert_link "Privacy", href: privacy_path
end
```

Author: Added a tip for it (`within :navigation, "Primary"`), keeping the plain `within :element` example for anyone not pulling in the gem. Do you think it's worth going into more detail here?

Reviewer 2: We try to avoid linking to gems aren't part of the default Gemfile or under the rails namespace.

Updating content in place is different, and welcome. A search field that filters
results as the user types, or a form that swaps its list of provinces when the
user picks a country, does not change the context: the user stays on the same
page, with the same focus and the same tab order. Only the fragment of content
that needs to change does.

### Disabling Controls Carefully

The `disabled` attribute marks a control as unavailable: the browser blocks
interaction and screen readers announce the disabled state. It is the right tool
when a control is genuinely unavailable, such as a feature behind a paid plan, a
payment method not offered in the user's country, or a field that does not apply
given another selection.

Two practices keep `disabled` usable:

* **Style the disabled state** so it stays readable without losing the disabled
  cue. Both [WCAG 1.4.3 Contrast (Minimum)][wcag-contrast] and [WCAG 1.4.11
  Non-text Contrast][wcag-non-text-contrast] exempt inactive controls, so
  browser defaults can render them at very low contrast. A small CSS rule
  overrides the default while still signaling the state:

  ```css
  :disabled {
    opacity: 0.6;
    cursor: default;
  }
  ```

> * **Explain why nearby**, in visible text or in the surrounding layout. A

Reviewer 2: "Explain why nearby" was a bit confusing upon first read.
Maybe expand it a bit for more context?
```suggestion
* **Explain why the control is disabled nearby**, in visible text or in the surrounding layout. A
```

single column, a right-to-left translation flips left and right, a mobile layout
collapses everything ([WCAG 1.3.3 Sensory
Characteristics][wcag-sensory-characteristics]). Refer to content by its name
and link to it where possible:

[wcag-sensory-characteristics]: https://www.w3.org/WAI/WCAG22/Understanding/sensory-characteristics.html

* Avoid: "See the form below." Prefer: "Refer to the **Sign-up form**" (linked
  when possible).
* Avoid: "Use the navigation on the left." Prefer: "Use the primary navigation."
* Avoid: "Click the button at the bottom right." Prefer: "**Save** the form."

NOTE: Responsive layouts that genuinely require a different visual order at
different breakpoints have no perfect solution: there is no way to ship one HTML
source that matches every breakpoint's reading order. Building mobile-first
reduces the risk in practice, since a source order that works on a single column
tends to translate cleanly to wider layouts. When a breakpoint truly cannot
avoid a reorder, test the result at that width with a keyboard and a screen
reader to confirm the experience is still understandable.

### Horizontal Scrolling

> Wide tables, image rails, and "card carousels" are often forced into a

Reviewer 2: I wasn't familiar with the term "image rails", it should probably be quoted like "card carousels"?
```suggestion
Wide tables, "image rails", and "card carousels" are often forced into a
```

1.4.1 Use of Color][wcag-use-of-color]):

```html+erb
<%# Avoid: only the red color of the label marks this field as required. %>
<%= form.label :email_address, class: "required" %>
<%= form.email_field :email_address %>

<%# Prefer: the word "required" sits alongside the visual cue,
    so the requirement also reaches users who cannot perceive the color. %>
<%= form.label :email_address, "Email address (required)" %>
<%= form.email_field :email_address, required: true %>
```

The same principle applies to links inside running text. A link that differs
from the surrounding text only in color is invisible to color-blind users and to
anyone reading in a context that strips color (printed pages, certain forced
palettes), so the underline carries the cue the color cannot. Italic or bold
treatment alone is technically allowed as the non-color cue, but in prose both
already signal other things (emphasis, citation), so a user has no way to tell
that a particular italic word is a link rather than an emphasized phrase.
Underline is the convention precisely because it does not collide with prose
semantics:

```css
p a {
  color: #0044cc;
  text-decoration: underline;
}
```

> TIP: The [axe browser extension](https://www.deque.com/axe/devtools/extension/),

Reviewer 2: This TIP feels like it should be moved before or after the section about declaring `color` and `background-color` together, not after using `text-decoration` for links.

Two CSS habits keep layouts resilient at any zoom level:

* **Use relative units (`rem`, `em`, `%`) for font sizes, paddings, widths, and
  media query breakpoints.** Absolute units like `px` ignore the user's
  preferred base font size, so a layout built on px values does not honor a
  reader's choice to enlarge default text in their browser or operating system.
  The [`ch`][ch-unit] unit (one `ch` equals the width of the `0` character in
  the current font) is especially useful for capping the width of prose: setting
  `max-width: 65ch` on `<p>`, `<li>`, or article bodies keeps lines at a
  comfortable reading length regardless of font size.
* **Let containers grow with their content.** Avoid fixed pixel heights on text
  containers, and allow wrapping rather than clipping overflowing text.

[ch-unit]: https://developer.mozilla.org/en-US/docs/Web/CSS/length#ch

Leave the root `font-size` alone, or set it to `100%` to make the inheritance
explicit, so the page picks up whatever the user configured in their browser or
operating system. Tricks that reset it for math convenience, like `html {
font-size: 62.5% }`, force every reader to bump their preferred size by 60

> percent just to land back on what they originally chose. Form controls escape
> that inheritance by default and ship with browser-fixed typography that ignores
> the body cascade, so add one rule to bring them back in line:
> 
> ```css
> select, textarea, input, button {
>   font: inherit;
> }
> ```

Reviewer 2: While it's a good CSS trick that I wasn't aware of, I'm wondering if we should just remove this as we just recommended not using `html { font-size: 62.5% }`
```suggestion
percent just to land back on what they originally chose. 
```

morphing:

```html+erb
<%= link_to "Email", team_members_path(sort_by: "email"), data: { turbo_action: "replace" } %>
```

With this setup, `autofocus` only moves focus on genuine navigations to a
different page, and interactions like sorting a table keep the user right where
they were.

NOTE: Turbo Drive can animate page transitions through the browser's [View
Transitions API][view-transitions] when the layout opts in with `<meta
name="view-transition" content="same-origin">`. Turbo checks
`prefers-reduced-motion` before running the animation and skips it when the user
has requested reduced motion, so the opt-in respects the same user preference
the rest of the styling should honor.

[view-transitions]: https://developer.mozilla.org/en-US/docs/Web/API/View_Transition_API

#### Turbo Frames

[Turbo Frames][turbo-frames] replace content within a frame. Lazy-loaded
sections, such as a comments thread that only loads when the user requests it,
are a common use case. These replacements do not move focus automatically, so
activating the link to show the content leaves the user with no signal that
anything happened.

> Placing `autofocus` inside the replacement markup directs focus to the revealed
> content once the frame updates. The [`turbo_frame_tag`][turbo_frame_tag] helper
> renders the frame:

Reviewer 2: The description of `turbo_frame_tag` put me off a bit here. It had me wonder what I should pay attention to.
Maybe explain the examples and describe `turbo_frame_tag` as a side note?
```suggestion
Placing `autofocus` inside the replacement markup directs focus to the revealed
content once the frame updates. For example, showing the comments for an article by using a Turbo Frame (the [`turbo_frame_tag`][turbo_frame_tag] helper
renders the frame):
```

#### Integrating axe-core in System Tests

New Rails applications no longer include system tests by default. Generate the
scaffolding first, which creates `test/application_system_test_case.rb` and a
`test/system/` directory:

```bash
$ bin/rails generate system_test accessibility_audits
```

See [Generating System Tests](testing.html#generating-system-tests) for details.
Then add the [`axe-core-capybara`][axe-core-capybara] integration and define an
`assert_accessible` helper on
[`ApplicationSystemTestCase`][ApplicationSystemTestCase] that forwards keyword
arguments to `Axe::Matchers::BeAxeClean`:

[axe-core-capybara]: https://github.com/dequelabs/axe-core-gems/tree/develop/packages/axe-core-capybara
[ApplicationSystemTestCase]: https://api.rubyonrails.org/classes/ActionDispatch/SystemTestCase.html

```ruby
# Gemfile
group :test do
  # ...
  gem "axe-core-capybara"
end
```

```ruby
# test/application_system_test_case.rb
require "test_helper"
require "axe-capybara"
require "axe/matchers/be_axe_clean"

class ApplicationSystemTestCase < ActionDispatch::SystemTestCase
  driven_by :selenium, using: :headless_chrome, screen_size: [1400, 1400]

  def assert_accessible(**options)

>     options
>       .inject(Axe::Matchers::BeAxeClean.new) do |matcher, (method, value)|
>         matcher.public_send(method, *Array.wrap(value))
>       end
>       .audit(page).then do |audit|
>         assert audit.passed?, audit.failure_message
>       end

Reviewer 2: We don't typically use leading dots in guides examples and chain methods calls after `end` on the same line:
```suggestion
    options.inject(Axe::Matchers::BeAxeClean.new) do |matcher, (method, value)|
      matcher.public_send(method, *Array.wrap(value))
    end.audit(page).then do |audit|
      assert audit.passed?, audit.failure_message
    end
```

  end
end
```

```ruby
# test/system/accessibility_audits_test.rb
require "application_system_test_case"

class AccessibilityAuditsTest < ApplicationSystemTestCase
  test "home page is accessible" do
    visit root_path
    assert_accessible
  end
end
```

Calling `assert_accessible` without options runs the rule set axe enables by
default: WCAG 2.0, 2.1, and 2.2 Level A and AA, plus best-practice rules.
AAA-level rules and experimental rules are excluded by default and must be opted
into explicitly.

> Each keyword argument maps to a chainable method on the matcher:

Reviewer 2: We call the "keyword argument" "option" above and below, maybe use it here as well?
```suggestion
Each option maps to a chainable method on the matcher:
```

Or maybe keep "argument":
```suggestion
Each option argument maps to a chainable method on the matcher:
```

end
```

`within_fieldset` scopes a query to a `<fieldset>` by its id or `<legend>`:

```ruby
within_fieldset "Billing cycle" do
  choose "Monthly"
end
```

Capybara's finders also accept filters for the same states assistive
technologies report: `focused`, `disabled`, `checked`, and ARIA attributes.
Prefer these filters over CSS pseudo-class selectors so the test reads like the
behavior it checks:

```ruby
# Avoid: CSS selector with a pseudo-class, brittle and harder to read
assert_css "input#user_email_address:focus"

# Prefer: semantic finder with a state filter
assert_field "Email address", focused: true
assert_field "Remember me", checked: true
```

For most flows, the semantic finders above are enough. Some interactions, such
as verifying tab order through a dialog or that a skip link lands on the main
content, depend on keyboard input specifically. Capybara drives those through
`send_keys` on the page or on an element:

```ruby
# Move focus to the next focusable element
page.send_keys :tab

# Activate the focused element
page.send_keys :enter
```

Use keyboard driving for the flows that depend on it, and the semantic finders
everywhere else.

> ### Manual Testing

Reviewer 2: This section goes into detail about using third party tools (how to install, which commands to use, menu items etc). I think that is a bit out of scope for this guide?
While I think it is useful having a central place to describe these tools, it can easily get outdated. Tools change and new tools get released.

Maybe just briefly describe the tools and link to their homepages for details?
